### PR TITLE
Remove time-series grouping widget for custom date breakout

### LINKED ISSUE #33504: Summarizing by custom date column does not allow date grouping
### Describe the bug

If I create a new custom column which is a date, I can summarize by that column, but selecting a grouping for that summary (by Day, by Month, by Year, by Hour) does nothing. It allows me to select it, but there is no change in the visualization nor in the generated SQL. 

### To Reproduce

1. Go to New Question
2. Select a table with a timestamp or date column
3. Create a new custom column that is just the value of the original date column. (Or use convertTimezone, or any formula you want. None seem to matter)
4. Add a count summarization with that new custom column. Click the arrow to select by Month, Day, etc. 
5. Refresh the visualization and you'll see it doesn't change

### Expected behavior

It should group by the date grouping selected

### Logs

_No response_

### Information about your Metabase installation

```JSON
Cloud Metabase
```

### Severity

annoying

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by WiNloSt**
Closes https://github.com/metabase/metabase/issues/33504

### Description

As Alex P. has pointed out in [this comment](https://github.com/metabase/metabase/issues/33504#issuecomment-1695571921), we shouldn't display the time-series grouping widget for custom date breakout because we don't support it.

This PR removes the time-series grouping widget when we summarize by custom date field. And ensure we won't display mode footer when the mode is outdated. e.g. when modifying a question without rerunning the result. Since `mode` relies on the last run result.

### How to verify

1. New > Question > Orders
1. Add a custom column `[Created At]` name it `Custom Created At`
1. Summarize count by `Custom Created At`
1. Click **Visualize**

You shouldn't see the time-series grouping widget in the footer.

#### After
Screenshot 2023-09-07 at 11 29 40 PM

#### Before
Screenshot 2023-09-07 at 11 30 01 PM

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase-types/api/query.ts [hunk 1]
```diff
@@ -355,7 +355,8 @@ export type Expression =
       ExpressionOperand,
       ExpressionOperand,
       ExpressionOperand,
-    ];
+    ]
+  | ConcreteFieldReference;
```
Comment (WiNloSt) [inline frontend/src/metabase-types/api/query.ts:359]: I think we're missing this type. As the screenshot below shows, an expression could be a field reference or a reference to another expression.

Screenshot 2023-09-07 at 11 17 25 PM

https://github.com/metabase/metabase/blob/a90511e342ea8b90455b7ead2ba3967780ef6f19/frontend/src/metabase-types/api/query.ts#L290-L297

I use it [in the test](https://github.com/metabase/metabase/pull/33810/files#diff-dbaaca4e4bf6f621873aa7622e48a9d5a7b1f7df0272354b12d7c2f406670a41R78-R80), otherwise, TypeScript would complain the type is incorrect.

File: frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx [hunk 2]
```diff
@@ -229,6 +260,43 @@ describe("QueryBuilder", () => {
...
         expect(screen.getByDisplayValue(TEST_CARD.name)).toBeInTheDocument();
       });
+
+      it("render time-series grouping widget for date field breakout", async () => {
+        await setup({
+          card: TEST_TIME_SERIES_WITH_DATE_BREAKOUT_CARD,
+        });
+        const timeSeriesModeFooter = await screen.findByTestId(
+          "timeseries-mode-bar",
+        );
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx:270]: It's getting the footer which is this component
Screenshot 2023-09-07 at 11 34 08 PM

File: frontend/src/metabase/visualizations/click-actions/components/TimeseriesGroupingWidget.jsx [hunk 3]
```diff
@@ -30,7 +30,9 @@ export class TimeseriesGroupingWidget extends Component {
       return (
         <PopoverWithTrigger
           triggerElement={
-            <SelectButton hasValue>{dimension.subDisplayName()}</SelectButton>
+            <SelectButton hasValue dataTestId="time-series-grouping">
```
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/click-actions/components/TimeseriesGroupingWidget.jsx:33]: Just added `dataTestId`

File: frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx [hunk 4]
```diff
@@ -229,6 +260,43 @@ describe("QueryBuilder", () => {
 
         expect(screen.getByDisplayValue(TEST_CARD.name)).toBeInTheDocument();
       });
+
+      it("render time-series grouping widget for date field breakout", async () => {
+        await setup({
+          card: TEST_TIME_SERIES_WITH_DATE_BREAKOUT_CARD,
+        });
+        const timeSeriesModeFooter = await screen.findByTestId(
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx:268]: nit:
- I'd call a "footer" the thing below this bar - [see screenshot](https://github.com/metabase/metabase/assets/6830683/cdadd931-a663-4e99-a81c-2e48b270f46e) ("footer" is something that is last, and there's nothing below it)
- ideally we match the `data-testid` name

```suggestion
        const timeSeriesModeBar = await screen.findByTestId(
```

**Review by kamilmielnik — APPROVED** [review]
LGTM :+1: 

Bonus points for covering both cases with tests (not only the problematic case) :100:

File: frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx (hunk 4 above)
Comment (WiNloSt) [reply frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx:268]: I was following [the component name](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/visualizations/click-actions/components/TimeseriesModeFooter.tsx), not the test ID. I'll make sure they are the same. The thing is that we've been using `ModeFooter` [for a long time](https://github.com/metabase/metabase/blob/d3ff5fdc5fca87d3f7e01a39c8890d599a927260/frontend/src/metabase/query_builder/components/view/View.jsx#L112) (At least 4 years), so I'll stick to that for now.

File: frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx [hunk 5]
```diff
@@ -229,6 +260,43 @@ describe("QueryBuilder", () => {
...
+          within(timeSeriesModeFooter).getByText("by"),
+        ).toBeInTheDocument();
+        expect(
+          within(timeSeriesModeFooter).getByTestId(
+            "time-series-grouping-select-button",
+          ),
+        ).toBeInTheDocument();
+      });
+
+      it("don't render time-series grouping widget for custom date field breakout (metabase#33504)", async () => {
```
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx]: fixed in cf9cdbd9c9fb4991aa63fd957c8440c257a6a1b7

File: frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx (hunk 4 above)
Comment (WiNloSt) [reply frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx:268]: Fixed in 882f7b9f05435386dc3e5012817d9f00a3a78351

File: frontend/src/metabase/query_builder/components/view/ConvertQueryModal/ConvertQueryModal.tsx [hunk 6]
```diff
@@ -10,6 +10,7 @@ import {
 } from "metabase/query_builder/selectors";
 import type { NativeQueryForm } from "metabase-types/api";
 import type { QueryBuilderUIControls, State } from "metabase-types/store";
+import type { UpdateQuestionOpts } from "metabase/query_builder/actions";
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/components/view/ConvertQueryModal/ConvertQueryModal.tsx:13]: Use the actual type instead of hard coding it.

File: frontend/src/metabase/query_builder/components/view/ConvertQueryModal/ConvertQueryModal.tsx [hunk 7]
```diff
@@ -56,7 +53,9 @@ const ConvertQueryModal = ({
     const newDatasetQuery = createDatasetQuery(query, question);
     const newQuestion = question.setDatasetQuery(newDatasetQuery);
 
-    onUpdateQuestion?.(newQuestion, { shouldUpdateUrl: true });
+    // We should rerun the question to ensure we get the proper `mode.ModeFooter` since
+    // `mode` relies on `qb.lastRunCard`
+    onUpdateQuestion?.(newQuestion, { run: true, shouldUpdateUrl: true });
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/components/view/ConvertQueryModal/ConvertQueryModal.tsx:58]: Without rerunning the question we'd still show the mode footer even if the question is already changed to `native` and not a `timeseries` question anymore.

This is because the `mode` selector was relying on the previous run card. So without rerunning the card, we still get `timeseries` mode even though the question has turned native.

Screenshot 2023-09-26 at 1 23 12 PM

File: frontend/src/metabase/visualizations/click-actions/components/TimeseriesModeFooter.tsx [hunk 8]
```diff
@@ -11,17 +11,27 @@ export const TimeseriesModeFooter = (props: Props): JSX.Element => {
     const { updateQuestion } = props;
     updateQuestion(question, { run: true });
   };
+  // The first breakout is always a date.
+  // See https://github.com/metabase/metabase/blob/e7363d97d6ed0ec8f5288a642e4990e85df57e79/frontend/src/metabase/visualizations/click-actions/Mode/utils.ts#L41-L44
+  const [dateBreakout] = props.query.breakouts();
+  const shouldHideTimeseriesGroupingWidget = dateBreakout
```
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/click-actions/components/TimeseriesModeFooter.tsx:17]: This is the fix.

File: e2e/test/scenarios/question/reproductions/28599-model-conversion-time-footer.cy.spec.js [hunk 9]
```diff
@@ -36,7 +36,7 @@ describe("issue 28599", () => {
   });
 
   it("should not show time granularity footer after question conversion to a model (metabase#28599)", () => {
-    cy.findByTestId("timeseries-mode-bar").within(() => {
+    cy.findByTestId("time-series-mode-footer").within(() => {
```
Comment (WiNloSt) [inline e2e/test/scenarios/question/reproductions/28599-model-conversion-time-footer.cy.spec.js:39]: Fix conflict from merging master.

File: frontend/src/metabase/query_builder/components/view/View.jsx [hunk 10]
```diff
@@ -382,7 +388,14 @@ class View extends Component {
...
 
-        {ModeFooter && (
+        {/* ModeFooter relies on `getMode` which relies on `qb.lastRunCard`,
+         so after a question modification, unless the result is refreshed,
+         we couldn't be sure if `mode` is still valid.
+         
+         We also couldn't just use `isResultDirty` because it doesn't consider
+         converting a question to a model as dirty.
+         */}
+        {ModeFooter && !isResultDirtyIgnoringModels && (
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/components/view/View.jsx:398]: The problem was that `isResultDirty` would still be `false` even if we just converted a question to a model.

But in this case of a footer, we want to make sure turning a question into a model is considered `dirty` since that changes the underlying query and [we don't really support models](https://github.com/metabase/metabase/blob/c47e63c34d9f37f01af91a76b0246a9075ec98ff/frontend/src/metabase/visualizations/click-actions/Mode/utils.ts#L15) when calculating mode in the first place.

File: frontend/src/metabase/query_builder/selectors.js [hunk 11]
```diff
@@ -459,6 +471,35 @@ export const getIsResultDirty = createSelector(
   },
 );
 
+export const getIsResultDirtyIgnoringModels = createSelector(
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/selectors.js:474]: This is basically `getIsResultDirtyIgnoringModels`, but in the end we don't call `areModelsEquivalent`.

File: frontend/src/metabase/query_builder/components/view/View.jsx (hunk 10 above)
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/view/View.jsx:398]: I don't understand how this change is connected to the original issue. Could you explain? If that's a separate issue please exclude it from the PR. It is very tricky and most likely incorrect.

**Review by ranquild — COMMENTED** [review]
Comments inline

File: frontend/src/metabase/query_builder/components/view/View.jsx (hunk 10 above)
Comment (WiNloSt) [reply frontend/src/metabase/query_builder/components/view/View.jsx:398]: @ranquild 
It's true that it's a separate issue, but the fix I made crashed a lot of tests because we're currently showing the footer when we're not supposed to. And the fix assumes we would be able to get the question breakout and it's dimension which won't be the case when we show the footer when we're not supposed to.

I could make the fix more flexible with a lot of guard clauses (on `TimeseriesModeFooter.tsx`) to account for when we can't get question breakouts and dimensions, but that doesn't feel like the right thing to do since we know we're not supposed to also render the footer in some cases anyway. Then create a new PR to fix the issue where we render the footer when we're not supposed to then come back to remove all the guard clauses on `TimeseriesModeFooter.tsx`.

I don't like the fact that `master` would be in an intermediate state where the fix needs to have a lot of guard clauses because we still render it when we're not supposed to, but if that's best then I would do that.

what do you think? Should I extract not rendering the footer into another PR?

Comment (ranquild) [reply frontend/src/metabase/query_builder/components/view/View.jsx:398]: Which tests exactly?

Comment (WiNloSt) [reply frontend/src/metabase/query_builder/components/view/View.jsx:398]: - `e2e/test/scenarios/native/reproductions/32121-sql-source-query-can-explore-results.cy.spec.js`
- `e2e/test/scenarios/question/reproductions/28599-model-conversion-time-footer.cy.spec.js`
- `e2e/test/scenarios/models/models-with-aggregation-and-breakout.cy.spec.js`
- `e2e/test/scenarios/native/native.cy.spec.js`
- `e2e/test/scenarios/visualizations/reproductions/21615-convert-to-sql.cy.spec.js`

These would be all I think.

Comment (WiNloSt) [reply frontend/src/metabase/query_builder/components/view/View.jsx:398]: The result can be seen from [this comment](https://github.com/metabase/metabase/pull/33810#issuecomment-1710507209) from deploysentinel and in its history

Comment (WiNloSt) [reply frontend/src/metabase/query_builder/components/view/View.jsx:398]: @ranquild This [was an example](https://www.deploysentinel.com/ci/runs/64fb31bdcc8487467c779669), where the test crashed. That run was from before `isResultDirtyIgnoringModels` is introduced.

Comment (ranquild) [reply frontend/src/metabase/query_builder/components/view/View.jsx:398]: @WiNloSt did you figure out why the app crashes?

Comment (WiNloSt) [reply frontend/src/metabase/query_builder/components/view/View.jsx:398]: Yeah, basically it's because of the new fix as stated.
> It's true that it's a separate issue, but the fix I made crashed a lot of tests because we're currently showing the footer when we're not supposed to. And the fix assumes we would be able to get the question breakout and it's dimension which won't be the case when we show the footer when we're not supposed to.

It's basically from this few lines of the fix
https://github.com/metabase/metabase/blob/c2bb9e95f165421d8310ff9e8e7d6707c350d310/frontend/src/metabase/visualizations/click-actions/components/TimeseriesModeFooter.tsx#L16-L19

The fix assumes that it'd only be rendered when the question is timeseries. However that's not always the case e.g. when modifying a GUI query without rerunning the result, turning GUI question native, turning a GUI question into a model.

> I could make the fix more flexible with a lot of guard clauses (on TimeseriesModeFooter.tsx) to account for when we can't get question breakouts and dimensions, but that doesn't feel like the right thing to do since we know we're not supposed to also render the footer in some cases anyway.

I could also do that but it looks wrong in my opinion since the timeseries footer shouldn't have to care if the question passed isn't timeseries (it should always get timeseries question), but I'll check with the product first as you suggested.

Comment (ranquild) [reply frontend/src/metabase/query_builder/components/view/View.jsx:398]: @WiNloSt take a look at these checks https://github.com/metabase/metabase/blob/31e83ed96dbe8d26b81a62db25ee841b5f6b163b/frontend/src/metabase/visualizations/click-actions/components/TimeseriesGroupingWidget.jsx#L22. I think we need to lift them up

Comment (ranquild) [reply frontend/src/metabase/query_builder/components/view/View.jsx:398]: E.g. don't render `TimeseriesGroupingWidget` if it isn't a MBQL question or there are no breakouts, or if there are no subdimensions https://github.com/metabase/metabase/blob/1f7e261ec4702b36aa1f815440396187c4b7e931/frontend/src/metabase/query_builder/components/TimeGroupingPopover.jsx#L27

**Comment by WiNloSt** [comment]
@ranquild I asked product [here](https://metaboat.slack.com/archives/C01LQQ2UW03/p1695727425016609)

File: frontend/src/metabase/query_builder/components/view/View.jsx (hunk 10 above)
Comment (WiNloSt) [reply frontend/src/metabase/query_builder/components/view/View.jsx:398]: @ranquild That would also work, but I think the real question is should `TimeseriesModeFooter.tsx` still need accept non-timeseries questions?

From these lines of code
https://github.com/metabase/metabase/blob/7e0eb68fd75fe69849dd263ce07dd2cb1aa67870/frontend/src/metabase/visualizations/click-actions/lib/modes.ts#L44-L45

It's where we determine what mode we should be using for the visualization. It feels to me that we should be able to expect that `TimeseriesMode.ModeFooter` would only get timeseries questions since it's been checked before.
https://github.com/metabase/metabase/blob/7e0eb68fd75fe69849dd263ce07dd2cb1aa67870/frontend/src/metabase/visualizations/click-actions/lib/modes.ts#L29

However, there is currently a problem with how this `getMode` function is being used in the selector.

https://github.com/metabase/metabase/blob/7e0eb68fd75fe69849dd263ce07dd2cb1aa67870/frontend/src/metabase/query_builder/selectors.js#L533-L538

The question is actually from the last run question, not the current question. That's why it's possible for the mode to be stale. e.g. when turning a question into a native question. We don't rerun the question, so `mode` would still be `timeseries` when it should have been `native` instead.

tl;dr is stale `mode` correct? I'm not sure about all the implications, but if it's correct then I could modify this fix so that the widget will account for non-timeseries questions passed to it.

File: frontend/src/metabase/visualizations/types/click-actions.ts [hunk 12]
```diff
@@ -167,5 +167,5 @@ export interface QueryClickActionsMode {
 
   clickActions: LegacyDrill[];
   fallback?: LegacyDrill;
-  ModeFooter?: (props: ModeFooterComponentProps) => JSX.Element;
+  ModeFooter?: (props: ModeFooterComponentProps) => JSX.Element | null;
```
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/types/click-actions.ts:170]: Since the type in `frontend/src/metabase/visualizations/click-actions/components/TimeseriesModeFooter.tsx` also has changed.

File: frontend/src/metabase/visualizations/click-actions/components/TimeseriesModeFooter.tsx [hunk 13]
```diff
@@ -1,27 +1,46 @@
...
+  if (!(props.query instanceof StructuredQuery)) {
+    return null;
+  }
+  const [breakout] = props.query.breakouts();
+  if (!breakout) {
+    return null;
+  }
+  const shouldHideTimeseriesGroupingWidget = breakout
+    .dimension()
+    .isExpression();
```
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/click-actions/components/TimeseriesModeFooter.tsx:28]: This is the fix for the bug https://github.com/metabase/metabase/pull/33810

Comment (ranquild) [reply frontend/src/metabase/visualizations/click-actions/components/TimeseriesModeFooter.tsx:28]: Could you check for the list of subdimensions instead? E.g. `dimension.dimensions()` to have more than 1 item

**Review by ranquild — COMMENTED** [review]
Looks good, one small comment about the dimensions check

File: frontend/src/metabase/visualizations/click-actions/components/TimeseriesModeFooter.tsx (hunk 13 above)
Comment (WiNloSt) [reply frontend/src/metabase/visualizations/click-actions/components/TimeseriesModeFooter.tsx:28]: That doesn't work in this particular case aka. custom date column since I can still see a lot of subdimension from the custom columns

Screenshot 2023-09-27 at 5 57 01 PM

I initially tried that and it also failed the unit tests I added in `frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx`, so that solution doesn't work for this specific case.

File: frontend/src/metabase/visualizations/click-actions/components/TimeseriesModeFooter.tsx [hunk 14]
```diff
@@ -1,27 +1,46 @@
...
+    <div className="flex layout-centered" data-testid="time-series-mode-footer">
       <span className="mr1">{t`View`}</span>
       <TimeseriesFilterWidget {...props} card={props.lastRunCard} />
-      <span className="mx1">{t`by`}</span>
-      <TimeseriesGroupingWidget
-        {...props}
-        onChange={onChange}
-        card={props.lastRunCard}
-      />
+      {!shouldHideTimeseriesGroupingWidget && (
```
Comment (ranquild) [inline frontend/src/metabase/visualizations/click-actions/components/TimeseriesModeFooter.tsx:34]: nit: double negation is hard to follow, maybe rename the var to something like `hasGroupingWidget = !breakout.dimension().isExpression();` and then `{hasGroupingWidget && (`

**Review by ranquild — APPROVED** [review]
